include/skipper/sequential_map.ipp

```diff
@@ -0,0 +1,238 @@
+}
+
+template <typename Key, typename Value>
+auto SequentialSkipListMap<Key, Value>::Iterator::operator->() const
+    -> const Element* {
+  return &ptr_->element;
+}
+
+template <typename Key, typename Value>
+auto SequentialSkipListMap<Key, Value>::Iterator::operator++(/* prefix */)
+    -> SequentialSkipListMap::Iterator& {
+  ptr_ = ptr_->Next();
+  return *this;
+}
+
+template <typename Key, typename Value>
+auto SequentialSkipListMap<Key, Value>::Iterator::operator++(int /* postfix */)
+    -> SequentialSkipListMap::Iterator {
+  auto copy = *this;
+  ++(*this);
+  return copy;
+}
+
+template <typename Key, typename Value>
+auto SequentialSkipListMap<Key, Value>::Iterator::operator==(
+    const SequentialSkipListMap::Iterator& other) const -> bool {
+  return ptr_ == other.ptr_;
+}
+
+template <typename Key, typename Value>
+auto SequentialSkipListMap<Key, Value>::Iterator::operator!=(
+    const SequentialSkipListMap::Iterator& other) const -> bool {
+  return !(*this == other);  // NOLINT (simplification will lead to recursion)
+}
+
+////////////////////////////////////////////////////////////////////////////////
+////
+//// SequentialSkipListMap: public interface
+////
+
+template <typename Key, typename Value>
+auto SequentialSkipListMap<Key, Value>::Find(const Key& key) const
+    -> SequentialSkipListMap::Iterator {
+  if (auto node = Traverse(key); node && node->element.key == key) {
+    return Iterator{node.get()};
+  } else {
+    return End();
+  }
+}
+
+template <typename Key, typename Value>
+auto SequentialSkipListMap<Key, Value>::Insert(const Key& key,
+                                               const Value& value)
+    -> std::pair<Iterator, bool> {
+  auto update = ForwardNodePtrs{kMaxLevel + 1};
+  auto node = Traverse(key, &update);
+
+  if (node && !(key < node->element.key)) {
+    return {Iterator{node.get()}, false};
+  }
+
+  auto node_level = GenerateRandomLevel();
+  if (node_level > level_) {
+    std::fill(std::begin(update) + level_ + 1,
+              std::begin(update) + node_level + 1, head_);
+    level_ = node_level;
+  }
+
+  auto new_node = std::make_shared<Node>(key, value, node_level);
+  for (auto level = Level{0}; level <= node_level; ++level) {
+    auto i = static_cast<std::size_t>(level);
+    new_node->forward[i] = std::exchange(update[i]->forward[i], new_node);
+  }
+
+  return {Iterator{new_node.get()}, true};
+}
+
+template <typename Key, typename Value>
+auto SequentialSkipListMap<Key, Value>::operator[](const Key& key) -> Value& {
+  if (auto node = Find(key); node != End()) {
+    return node->value;
+  } else {
+    auto p = Insert(key, Value{});
+    return p.first->value;
+  }
+}
+
+template <typename Key, typename Value>
+auto SequentialSkipListMap<Key, Value>::operator[](const Key& key) const
+    -> const Value& {
+  if (auto node = Find(key); node != End()) {
+    return node->value;
+  } else {
+    throw std::invalid_argument("Given key does not exist.");
```

Comment:
Is this consistent with `std::map`?

Reply:
Do you mean the interface itself?

Reply:
No, just asking about the type of exception. I've just noticed that `std::map` doesn't have `operator[](Key) const`.
Should we remove it?